catalyst/data/sampler_inbatch.py

```diff
@@ -200,3 +241,164 @@ def _sample_from_distmat(
+
+        Args:
+            labels: labels of the batch, shape (batch_size,)
+
+        Returns:
+            matrix of indices of classes in batch
+        """
+        unique_labels = sorted(np.unique(labels))
+        labels_number = len(unique_labels)
+        labels_mask = torch.zeros(size=(labels_number, len(labels)))
+        for label_idx, label in enumerate(unique_labels):
+            label_indices = find_value_ids(labels, label)
+            labels_mask[label_idx][label_indices] = 1
+        return labels_mask.type(torch.bool)
+
+    @staticmethod
+    def _count_intra_class_distances(
+        embeddings: Tensor, mean_vectors: Tensor
+    ) -> Tensor:
+        """
+        Count matrix of distances from mean vector of each class to it's
+        samples embeddings.
+
+        Args:
+            embeddings: tensor of shape (p, k, embed_dim) where p is a number
+                of classes in the batch, k is a number of samples for each
+                class
+            mean_vectors: tensor of shape (p, embed_dim) -- mean vectors
+                of each class in the batch
+
+        Returns:
+            tensor of shape (p, k) -- matrix of distances from mean vectors to
+                related samples in the batch
+        """
+        p, k, embed_dim = embeddings.shape
+        # Create (p, k, embed_dim) tensor of mean vectors for each class
+        mean_vectors = mean_vectors.unsqueeze(1).repeat((1, k, 1))
+        # Count euclidean distance between embeddings and mean vectors
+        distances = torch.pow(embeddings - mean_vectors, 2).sum(2)
+        return distances
+
+    @staticmethod
+    def _count_inter_class_distances(mean_vectors: Tensor) -> Tensor:
+        """
+        Count matrix of distances from mean vectors of classes to each other
+
+        Args:
+            mean_vectors: tensor of shape (p, embed_dim) -- mean vectors
+                of classes
+
+        Returns:
+            tensor of shape (p, p) -- matrix of distances between mean vectors
+        """
+        distance = euclidean_distance(mean_vectors, mean_vectors)
+        return distance
+
+    @staticmethod
+    def _fill_diagonal(matrix: Tensor, value: float) -> Tensor:
+        """
+        Set diagonal elements with the value.
+
+        Args:
+            matrix: tensor of shape (p, p)
+            value: value that diagonal should be filled with
+
+        Returns:
+            modified matrix with inf on diagonal
+        """
+        p, _ = matrix.shape
+        indices = torch.diag(torch.ones(p)).type(torch.bool)
+        matrix[indices] = value
+        return matrix
+
+    def sample(self, features: Tensor, labels: TLabels) -> TTriplets:
+        """
+        This method selects the hardest positive and negative example for
+        each label in the batch. It counts mean vectors for all the labels
+        and choose the hardest positive sample from the batch and the hardest
+        mean vector for it.
```

Comment:
In comparison with docs from the top of the current class this one looks confused / ambiguous.
I offer to left only 1st exploration from the top of the class, it is really great.

```
    This sampler selects hardest triplets based on distance to mean vectors:
    anchor is a mean vector of features of i-th class in the batch,
    the hardest positive sample is the most distant from anchor sample of
    anchor's class, the hardest negative sample is the closest mean vector
    of another classes.
    The batch must contain k samples for p classes in it (k > 1, p > 1).
``` 
cc

Reply:
You can put this 1st peace of documentation here (in sample method) or in the top of the class.